ffi/examples/read-table/schema.h

```diff
@@ -0,0 +1,259 @@
+#include "delta_kernel_ffi.h"
+
+/**
+ * This module defines a very simple model of a schema, used only to be able to print the schema of
+ * a table. It consists of a "SchemaBuilder" which is our user data that gets passed into each visit_x
+ * call. This simply keeps track of all the lists we are asked to allocate.
+ *
+ * Each list is a "SchemaItemList", which tracks its length an an array of "SchemaItem"s.
+ *
+ * Each "SchemaItem" has a name and a type, which are just strings. It can also have a list which is
+ * its children. This is initially always NULL, but when visiting a struct, map, or array, we point
+ * this at the list specified in the callback, which allows us to traverse the schema when printing
+ * it. Note that this points to one of the lists in the builder's set of lists and is not a copy.
+ */
+
+// If you want the visitor to print out what it's being asked to do at each step, uncomment the
+// following line
+// #define PRINT_VISITS
+
+typedef struct SchemaItemList SchemaItemList;
+
+typedef struct {
+  char* name;
+  char* type;
+  SchemaItemList* children;
+} SchemaItem;
+
+typedef struct SchemaItemList {
+  uint32_t len;
+  SchemaItem* list;
+} SchemaItemList;
+
+typedef struct {
+  int list_count;
+  SchemaItemList* lists;
+} SchemaBuilder;
+
+char* allocate_name(const KernelStringSlice slice) {
+  char* buf = malloc(sizeof(char) * (slice.len + 1)); // +1 for null
+  snprintf(buf, slice.len + 1, "%s", slice.ptr);
+  return buf;
```

Comment:
Seems they've added it most places: https://www.gnu.org/software/gnulib/manual/html_node/strndup.html

I'll use it for now and people can complain, since it's definitely nicer.